File: KLASA_3/ROZDZIAL_1/LEKCJA_1/PYTHON/ZAD5.py
from stos import Stos


def czy_para(otwierajacy, zamykajacy):
    pary = {
        '(': ')',
        '{': '}',
        '[': ']'
    }
    return pary[otwierajacy] == zamykajacy


def sprawdz_wyrazenie(wyrazenie_nawiasowe):
    stos = Stos()
    for nawias in wyrazenie_nawiasowe:
        if nawias in '({[':
            stos.push(nawias)
        elif nawias in ')}]':
            if stos.isEmpty():
                return False
            else:
                otwierajacy = stos.peek()
                stos.pop()
                if not czy_para(otwierajacy, nawias):
                    return False
    return stos.isEmpty()


print('TAK' if sprawdz_wyrazenie(input('Podaj wyrażenie nawiasowe: ')) else 'NIE')

File: KLASA_3/ROZDZIAL_1/LEKCJA_1/C++/ZAD5.cpp
#include <iostream>
#include <stack>
#include <map>

using namespace std;

bool czy_para(char otwierajacy, char zamykajacy)
{
    map<char,char> pary;
    pary['('] = ')';
    pary['['] = ']';
    pary['{'] = '}';

    return pary[otwierajacy] == zamykajacy;
}

bool sprawdz_wyrazenie(string wyrazenie)
{
    stack<char> stos;

    for(char nawias: wyrazenie){
        if(nawias=='(' || nawias=='[' || nawias=='{'){
            stos.push(nawias);
        }else{
            if(stos.empty())return false;
            char otwierajacy = stos.top();
            stos.pop();
            if(!czy_para(otwierajacy, nawias))return false;
        }
    }

    return stos.empty();
}

int main()
{
    string wyrazenie;

    cout << "Podaj wyrazenie nawiasowe: ";
    cin >> wyrazenie;

    cout << (sprawdz_wyrazenie(wyrazenie) ? "TAK" : "NIE");

    return 0;
}
